libs/ml/tests/unit/serializers/serialisation.cpp

```diff
@@ -0,0 +1,4407 @@
+
+TYPED_TEST(SaveParamsTest, divide_saveparams_test)
+{
+  using TensorType    = TypeParam;
+  using DataType      = typename TypeParam::Type;
+  using VecTensorType = typename fetch::ml::ops::Ops<TensorType>::VecTensorType;
+  using SPType        = typename fetch::ml::ops::Divide<TensorType>::SPType;
+  using OpKind        = fetch::ml::ops::Divide<TensorType>;
+
+  TensorType data_1 = TensorType::FromString(
+      "1, -2, 3,-4, 5,-6, 7,-8;"
+      "1,  2, 3, 4, 5, 6, 7, 8");
+
+  TensorType data_2 = TensorType::FromString(
+      " 8, -7, 6,-5, 4,-3, 2,-1;"
+      "-8,  7,-6, 5,-4, 3,-2, 1");
+
+  OpKind op;
+
+  TensorType    prediction(op.ComputeOutputShape(
+      {std::make_shared<const TensorType>(data_1), std::make_shared<const TensorType>(data_2)}));
+  VecTensorType vec_data(
+      {std::make_shared<const TensorType>(data_1), std::make_shared<const TensorType>(data_2)});
+
+  op.Forward(vec_data, prediction);
+
+  auto   dsp = SerialiseDeserialiseBuild<SPType, TypeParam>(op);
+  OpKind new_op(*dsp);
+
+  // check that new predictions match the old
+  TensorType new_prediction(op.ComputeOutputShape(
+      {std::make_shared<const TensorType>(data_1), std::make_shared<const TensorType>(data_2)}));
+  new_op.Forward(vec_data, new_prediction);
+
+  // test correct values
+  EXPECT_TRUE(new_prediction.AllClose(prediction, DataType{0}, DataType{0}));
+}
+
+TYPED_TEST(SaveParamsTest, divide_saveparams_backward_test)
+{
+  using TensorType = TypeParam;
+  using OpKind     = fetch::ml::ops::Divide<TensorType>;
+  using SPType     = typename OpKind::SPType;
+
+  TensorType data_1 = TensorType::FromString(
+      "1, -2, 3,-4, 5,-6, 7,-8;"
+      "1,  2, 3, 4, 5, 6, 7, 8");
+
+  TensorType data_2 = TensorType::FromString(
+      "8, -7, 6,-5, 4,-3, 2,-1;"
+      "8,  7,-6, 5,-4, 3,-2, 1");
+
+  TensorType error = TensorType::FromString(
+      "1, -1, 2, -2, 3, -3, 4, -4;"
+      "5, -5, 6, -6, 7, -7, 8, -8");
+
+  fetch::ml::ops::Divide<TensorType> op;
+  std::vector<TensorType>            prediction = op.Backward(
+      {std::make_shared<TensorType>(data_1), std::make_shared<TensorType>(data_2)}, error);
+
+  auto   dsp = SerialiseDeserialiseBuild<SPType, TypeParam>(op);
+  OpKind new_op(*dsp);
+
+  // check that new predictions match the old
+  std::vector<TensorType> new_prediction = new_op.Backward(
+      {std::make_shared<TensorType>(data_1), std::make_shared<TensorType>(data_2)}, error);
+
+  // test correct values
+  EXPECT_TRUE(prediction.at(0).AllClose(
+      new_prediction.at(0), fetch::math::function_tolerance<typename TypeParam::Type>(),
+      fetch::math::function_tolerance<typename TypeParam::Type>()));
+  EXPECT_TRUE(prediction.at(1).AllClose(
+      new_prediction.at(1), fetch::math::function_tolerance<typename TypeParam::Type>(),
+      fetch::math::function_tolerance<typename TypeParam::Type>()));
+}
+
+//////////////////
+/// EMBEDDINGS ///
+//////////////////
+
+TYPED_TEST(SaveParamsTest, embeddings_saveparams_test)
```

Comment:
Do these two tests (for Embeddings I mean) use graph_builder?

Reply:
doesn't look like it; why are you asking? There are some tests in here that use BuildGraph and some that do not...